sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/types/PhysicalDataType.scala

```diff
@@ -17,53 +17,237 @@
+import org.apache.spark.unsafe.types.{ByteArray, UTF8String}
+
+sealed abstract class PhysicalDataType {
+  private[sql] type InternalType
+  private[sql] def ordering: Ordering[InternalType]
+  private[sql] val tag: TypeTag[InternalType]
+}
+
+object PhysicalDataType {
+  def apply(dt: DataType): PhysicalDataType = dt match {
+    case NullType => PhysicalNullType
+    case ByteType => PhysicalByteType
+    case ShortType => PhysicalShortType
+    case IntegerType => PhysicalIntegerType
+    case LongType => PhysicalLongType
+    case StringType => PhysicalStringType
+    case FloatType => PhysicalFloatType
+    case DoubleType => PhysicalDoubleType
+    case DecimalType.Fixed(p, s) => PhysicalDecimalType(p, s)
+    case BooleanType => PhysicalBooleanType
+    case BinaryType => PhysicalBinaryType
+    case TimestampType => PhysicalLongType
+    case TimestampNTZType => PhysicalLongType
+    case DayTimeIntervalType(_, _) => PhysicalLongType
+    case YearMonthIntervalType(_, _) => PhysicalIntegerType
+    case DateType => PhysicalIntegerType
+    case ArrayType(elementType, containsNull) => PhysicalArrayType(elementType, containsNull)
+    case StructType(fields) => PhysicalStructType(fields)
+    case MapType(keyType, valueType, valueContainsNull) =>
+      PhysicalMapType(keyType, valueType, valueContainsNull)
+    case _ => UninitializedPhysicalType
+  }
+
+  def ordering(dt: DataType): Ordering[Any] = apply(dt).ordering.asInstanceOf[Ordering[Any]]
+}
+
+trait PhysicalPrimitiveType
+
+class PhysicalBinaryType() extends PhysicalDataType {
+  private[sql] val ordering =
+    (x: Array[Byte], y: Array[Byte]) => ByteArray.compareBinary(x, y)
+
+  private[sql] type InternalType = Array[Byte]
+  @transient private[sql] lazy val tag = typeTag[InternalType]
+}
 case object PhysicalBinaryType extends PhysicalBinaryType
 
-class PhysicalBooleanType() extends PhysicalPrimitiveType
-case object PhysicalBooleanType extends PhysicalBooleanType
-
-class PhysicalByteType() extends PhysicalPrimitiveType
-case object PhysicalByteType extends PhysicalByteType
-
-class PhysicalCalendarIntervalType() extends PhysicalDataType
+class PhysicalBooleanType extends PhysicalDataType {
+  // The companion object and this class is separated so the companion object also subclasses
+  // this type. Otherwise, the companion object would be of type "BooleanType$" in byte code.
+  // Defined with a private constructor so the companion object is the only possible instantiation.
+  private[sql] type InternalType = Boolean
+  private[sql] val ordering = implicitly[Ordering[InternalType]]
+  @transient private[sql] lazy val tag = typeTag[InternalType]
+}
+case object PhysicalBooleanType extends PhysicalBooleanType with PhysicalPrimitiveType
```

Comment:
this is a bit different from the logical type. Shall we make the `class PhysicalBooleanType` extends `PhysicalPrimitiveType`?

Reply:
I believe there is no primitive type in logical data type? 

For example `class BooleanType private() extends AtomicType {`

Reply:
The reason I have PhysicalDataType implementation looks like this is I am trying to match PhysicalDataType with LogicalDataType classes. In the future we will need to deal with NumericType, IntegralType, etc. The PrimitiveType is not in LogicalDataType classes thus probably we cannot extend it in PhysicalDataType.

Reply:
````
case object BooleanType extends BooleanType
```

The case object for logical type is very simple. Why the case object of physical type needs to mix in extra traits?

Reply:
Because this PR introduced the PhysicalPrimitiveType `https://github.com/apache/spark/pull/40400`. The current way is to make that PR still works while we refactor physical data types.

Reply:
I feel risky if the `class BooleanType` doesn't extend `PhysicalPrimitiveType` while the previous code did. We are changing the class hierarchy here.

> one class can only extend one class

This is not true as `PhysicalPrimitiveType` is a trait

Reply:
Sure I will update the BooleanType